File: src/bollettino_neve.c
/* interface: bollettino_neve    server    porta */

#define _POSIX_C_SOURCE 200809L

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <unitypes.h>
#include <unistr.h>
#include <unistdio.h>
#include <sys/socket.h>
#include "../include/utils.h"
#include "../include/csexp.h"
#include "../include/extraio.h"

int main (int argc, char** argv){
    int sd, err;
    struct addrinfo hints, *res;
    char *usrIn[2]={'\0'};
    uint8_t *netIn;
    uint8_t *netOut;
    uint8_t *message[1]={0};

    usrIn[0]=malloc(MAX_LEN);
    usrIn[1]=malloc(MAX_LEN);

    /* Controllo parametri */
    if (argc != 3){
        perror ("Usage: bollettino_neve server porta\n");
        exit(EXIT_FAILURE);
    }
    if(is_port(argv[2]) < 0){
        perror ("port have to be number of a service port\n");
        exit (EXIT_FAILURE);
    }

    /* Direttive per getaddrinfo */
    memset(&hints, 0, sizeof(hints));
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;

    /* Risoluzione nome */
    if((err = getaddrinfo (argv[1],argv[2], &hints, &res)) != 0){
        fprintf(stderr,"Naming error: %s\n", gai_strerror(err));
        exit(EXIT_FAILURE);
    }

    /* Connessione con fallback */
    if ((sd=socket_fallback(res))<0){
        perror("Socket creation or connection error\n");
        exit(EXIT_FAILURE);
    }

    freeaddrinfo(res);

    /*** CONNESSIONE AVVENUTA CON SUCCESSO ***/

    /*** Interfaccia utente ***/
    printf("Inserire la regione di interesse\n");
    scanf("%s", usrIn[0]);
    while (strcmp(usrIn[0],"fine")){
        printf("Inserire il numero di località da visualizzare\n");
        scanf("%s", usrIn[1]);

        /* creazione messaggio */
        netOut = createmex_c(usrIn, 2);

        /* invio messaggio */
        if(write_all(sd, netOut, u8_strlen(netOut))<0){
            perror("write_all");
            exit (EXIT_FAILURE);
        }

        /* lettura messaggio dal server */
        if((err=readmex(sd, &netIn))<0){
            fprintf (stderr, "readmex error: %s\n", rm_getmessage(err));
            exit (EXIT_FAILURE);
        }
        if(getmexinfo(message, 1, netIn)<0){
            fprintf(stderr, "getmexinfo error\n");
            exit (EXIT_FAILURE);
        }

        ulc_fprintf (stdout, "%U\n", message[0]);

        /* new user request */
        printf("Inserire la regione di interesse\n");
        scanf("%s", usrIn[0]);
    }


    if (close(sd)<0){
        perror ("close");
        exit (EXIT_FAILURE);
    }
}

File: java_client/BollettinoNeve.java
// java implementation of client

import java.io.*;
import java.net.*;

public class BollettinoNeve {
    public static void main (String args[]){
        // main parameter control
        if (args.length != 2){
            System.out.println("Usage: java BollettinoNeve host port");
            System.exit(-1);
        }
        try {
            int portNum = Integer.parseInt(args[1]);
            if (portNum > 65535) new NumberFormatException("Invalid port");
        } catch (NumberFormatException e){
            System.out.println ("port parameter have to be a number of a valid service port\n");
            System.out.println (e.getMessage());
            System.exit (-1);
        }

        try{
            BufferedReader usrIn = new BufferedReader (new InputStreamReader(System.in));
            Socket s = new Socket (args[0], Integer.parseInt(args[1]));
            CSExpReader netIn = new CSExpReader(s.getInputStream());
            CSExpWriter netOut = new CSExpWriter(s.getOutputStream());

            // User interface
            String info[]= new String[2];
            System.out.println ("Inserire la regione di interesse:");
            info[0]=usrIn.readLine();
            while (!info[0].equals("fine")){
                System.out.println ("Inserire numero di località da visualizzare:");
                info[1]=usrIn.readLine();

                netOut.writemex(info);
                netOut.flush();

                // print all return messages
                String[] messages = netIn.readmex();
                System.out.println (messages[0] + "media: " + messages[1] + " cm");

                // New user request
                System.out.println ("\nInserire la regione di interesse");
                info[0]=usrIn.readLine();
            }
            
            s.close();
        } catch (Exception e){
            System.out.println (e.getMessage());
            e.getStackTrace();
            System.exit(-1);
        }
    }  
}
